File: core/src/main/java/com/github/fburato/highwheelmodules/core/algorithms/ModuleGraphTransitiveClosure.java
package com.github.fburato.highwheelmodules.core.algorithms;

import com.github.fburato.highwheelmodules.model.modules.HWModule;
import com.github.fburato.highwheelmodules.model.modules.ModuleDependency;
import com.github.fburato.highwheelmodules.model.modules.ModuleGraph;

import java.util.*;

/**
 * Calculate the Transitive closure of a ModuleGraph using the Floyd-Warshall algorithm Ref Cormen, Thomas H.;
 * Leiserson, Charles E.; Rivest, Ronald L. (1990). Introduction to Algorithms (1st ed.). MIT Press and McGraw-Hill.
 * ISBN 0-262-03141-8. See in particular Section 26.2, "The Floyd–Warshall algorithm" pp. 558–565
 */
public class ModuleGraphTransitiveClosure {

    public static class Difference {
        public final HWModule source;
        public final HWModule dest;
        public final int firstDistance;
        public final int secondDistance;

        public Difference(HWModule source, HWModule dest, int firstDistance, int secondDistance) {
            this.source = source;
            this.dest = dest;
            this.firstDistance = firstDistance;
            this.secondDistance = secondDistance;
        }

        @Override
        public String toString() {
            return "Difference{" + "source=" + source + ", dest=" + dest + ", firstDistance=" + firstDistance
                    + ", secondDistance=" + secondDistance + '}';
        }
    }

    public static class PathDifference {
        public final HWModule source;
        public final HWModule dest;
        public final List<HWModule> firstPath;
        public final List<HWModule> secondPath;

        public PathDifference(HWModule source, HWModule dest, List<HWModule> firstPath, List<HWModule> secondPath) {
            this.source = source;
            this.dest = dest;
            this.firstPath = firstPath;
            this.secondPath = secondPath;
        }

        @Override
        public String toString() {
            return "PathDifference{" + "source=" + source + ", dest=" + dest + ", firstPath=" + firstPath
                    + ", secondPath=" + secondPath + '}';
        }
    }

    private final List<HWModule>[][] minimumPathMatrix;
    private final Map<HWModule, Integer> indexMap;
    private final Collection<HWModule> modules;

    public ModuleGraphTransitiveClosure(ModuleGraph<ModuleDependency> moduleGraph, Collection<HWModule> modules) {
        this.modules = modules;
        minimumPathMatrix = initialiseSquareMatrixTo(modules.size());
        indexMap = createMapModuleIndex(modules);

        initialiseDistanceOneModules(modules, moduleGraph);

        applyFloydWarshallMainIteration();
    }

    @SuppressWarnings({ "unchecked", "rawtypes" })
    private List<HWModule>[][] initialiseSquareMatrixTo(final int size) {
        final List<HWModule>[][] array = (List<HWModule>[][]) new List[size][size];
        for (int i = 0; i < array.length; ++i)
            for (int j = 0; j < array.length; ++j)
                array[i][j] = new ArrayList<>();
        return array;
    }

    private Map<HWModule, Integer> createMapModuleIndex(Collection<HWModule> modules) {
        final Map<HWModule, Integer> map = new HashMap<>(modules.size());
        int moduleCount = 0;
        for (HWModule module : modules) {
            map.put(module, moduleCount++);
        }
        return map;
    }

    private void initialiseDistanceOneModules(Collection<HWModule> modules, ModuleGraph<ModuleDependency> moduleGraph) {
        for (HWModule module : modules) {
            for (HWModule dependency : moduleGraph.dependencies(module)) {
                minimumPathMatrix[indexMap.get(module)][indexMap.get(dependency)].add(dependency);
            }
        }
    }

    private void applyFloydWarshallMainIteration() {
        for (int i = 0; i < minimumPathMatrix.length; ++i) {
            for (int j = 0; j < minimumPathMatrix.length; ++j) {
                for (int k = 0; k < minimumPathMatrix.length; ++k) {
                    List<HWModule> pathIJ = minimumPathMatrix[i][j];
                    List<HWModule> pathIK = minimumPathMatrix[i][k];
                    List<HWModule> pathKJ = minimumPathMatrix[k][j];
                    if (pathIJ.isEmpty() && !pathIK.isEmpty() && !pathKJ.isEmpty()) {
                        minimumPathMatrix[i][j].clear();
                        minimumPathMatrix[i][j].addAll(pathIK);
                        minimumPathMatrix[i][j].addAll(pathKJ);
                    } else if (!pathIJ.isEmpty() && !pathIK.isEmpty() && !pathKJ.isEmpty()) {
                        if (pathIK.size() + pathKJ.size() < pathIJ.size()) {
                            minimumPathMatrix[i][j].clear();
                            minimumPathMatrix[i][j].addAll(pathIK);
                            minimumPathMatrix[i][j].addAll(pathKJ);
                        }
                    }
                }
            }
        }
    }

    public Boolean isReachable(HWModule vertex1, HWModule vertex2) {
        return minimumDistance(vertex1, vertex2).map((a) -> a < Integer.MAX_VALUE).orElse(false);
    }

    public boolean same(ModuleGraphTransitiveClosure other) {
        return diff(other).map(List::isEmpty).orElse(false);
    }

    public Optional<List<Difference>> diff(ModuleGraphTransitiveClosure other) {
        return diffPath(other).map((argument) -> {
            final List<Difference> result = new ArrayList<>(argument.size());
            for (PathDifference pathDifference : argument) {
                result.add(new Difference(pathDifference.source, pathDifference.dest, pathDifference.firstPath.size(),
                        pathDifference.secondPath.size()));
            }
            return result;
        });
    }

    public Optional<List<PathDifference>> diffPath(ModuleGraphTransitiveClosure other) {
        if (!modules.containsAll(other.modules) || !other.modules.containsAll(modules))
            return Optional.empty();
        final List<PathDifference> differences = new ArrayList<>();
        for (HWModule i : modules) {
            for (HWModule j : modules) {
                int thisI = indexMap.get(i), thisJ = indexMap.get(j), otherI = indexMap.get(i),
                        otherJ = indexMap.get(j);
                if (minimumPathMatrix[thisI][thisJ].size() != other.minimumPathMatrix[otherI][otherJ].size())
                    differences.add(new PathDifference(i, j, minimumPathMatrix[thisI][thisJ],
                            other.minimumPathMatrix[otherI][otherJ]));
            }
        }
        return Optional.of(differences);
    }

    public Optional<Integer> minimumDistance(HWModule vertex1, HWModule vertex2) {
        if (indexMap.get(vertex1) == null || indexMap.get(vertex2) == null)
            return Optional.empty();
        else {
            final int distance = minimumPathMatrix[indexMap.get(vertex1)][indexMap.get(vertex2)].size();
            return Optional.of(distance == 0 ? Integer.MAX_VALUE : distance);
        }
    }

    public List<HWModule> minimumDistancePath(HWModule vertex1, HWModule vertex2) {
        if (indexMap.get(vertex1) == null || indexMap.get(vertex2) == null)
            return new ArrayList<>();
        else
            return minimumPathMatrix[indexMap.get(vertex1)][indexMap.get(vertex2)];
    }
}

File: core/src/main/scala/com/github/fburato/highwheelmodules/core/algorithms/ModuleGraphTransitiveClosure.scala
package com.github.fburato.highwheelmodules.core.algorithms


import com.github.fburato.highwheelmodules.core.algorithms.ModuleGraphTransitiveClosure.{Difference, PathDifference}
import com.github.fburato.highwheelmodules.model.modules.{HWModule, ModuleDependency, ModuleGraph}

import scala.collection.mutable.ArrayBuffer

/**
 * Calculate the Transitive closure of a ModuleGraph using the Floyd-Warshall algorithm Ref Cormen, Thomas H.;
 * Leiserson, Charles E.; Rivest, Ronald L. (1990). Introduction to Algorithms (1st ed.). MIT Press and McGraw-Hill.
 * ISBN 0-262-03141-8. See in particular Section 26.2, "The Floyd–Warshall algorithm" pp. 558–565
 */
class ModuleGraphTransitiveClosure(
                                    private val minimumPathMatrix: Array[Array[ArrayBuffer[HWModule]]],
                                    private val indexMap: Map[HWModule, Int],
                                    private val modules: Seq[HWModule]
                                  ) {
  def isReachable(vertex1: HWModule, vertex2: HWModule): Boolean =
    minimumDistance(vertex1, vertex2).exists(d => d < Int.MaxValue)

  def same(other: ModuleGraphTransitiveClosure): Boolean =
    diff(other).exists(s => s.isEmpty)

  def diff(other: ModuleGraphTransitiveClosure): Option[Seq[ModuleGraphTransitiveClosure.Difference]] =
    for {
      pathDiff <- diffPath(other)
      transformed = pathDiff.map(pDiff => Difference(pDiff.source, pDiff.dest, pDiff.firstPath.size, pDiff.secondPath.size))
    } yield transformed

  def diffPath(other: ModuleGraphTransitiveClosure): Option[Seq[ModuleGraphTransitiveClosure.PathDifference]] = {
    def computePathDiff: Seq[ModuleGraphTransitiveClosure.PathDifference] =
      for {
        i <- modules
        j <- modules
        thisI = indexMap(i)
        thisJ = indexMap(j)
        otherI = other.indexMap(i)
        otherJ = other.indexMap(j)
        res = PathDifference(i, j, minimumPathMatrix(thisI)(thisJ).toSeq, other.minimumPathMatrix(otherI)(otherJ).toSeq)
        if minimumPathMatrix(thisI)(thisJ).size != other.minimumPathMatrix(otherI)(otherJ).size
      } yield res

    if (!modules.forall(m => other.modules.contains(m)) || !other.modules.forall(m => modules.contains(m))) {
      None
    } else {
      Some(computePathDiff)
    }
  }

  def minimumDistance(vertex1: HWModule, vertex2: HWModule): Option[Int] =
    for {
      v1Index <- indexMap.get(vertex1)
      v2Index <- indexMap.get(vertex2)
      distance = minimumPathMatrix(v1Index)(v2Index).size
    } yield if (distance == 0) Int.MaxValue else distance

  def minimumDistancePath(vertex1: HWModule, vertex2: HWModule): Seq[HWModule] =
    (for {
      v1Index <- indexMap.get(vertex1)
      v2Index <- indexMap.get(vertex2)
    } yield minimumPathMatrix(v1Index)(v2Index).toSeq).getOrElse(Seq())
}

object ModuleGraphTransitiveClosure {

  case class Difference(source: HWModule, dest: HWModule, firstDistance: Int, secondDistance: Int)

  case class PathDifference(source: HWModule, dest: HWModule, firstPath: Seq[HWModule], secondPath: Seq[HWModule])

  def apply(moduleGraph: ModuleGraph[ModuleDependency], modulesSeq: Seq[HWModule]): ModuleGraphTransitiveClosure = {
    val minimumPathMatrix = Array.ofDim[ArrayBuffer[HWModule]](modulesSeq.size, modulesSeq.size)
    for {
      i <- modulesSeq.indices
      j <- modulesSeq.indices
    } {
      minimumPathMatrix(i)(j) = new ArrayBuffer[HWModule]()
    }
    val mapModuleIndex = modulesSeq.zipWithIndex.toMap
    for {
      m <- modulesSeq
      dependency <- moduleGraph.dependencies(m)
    } {
      minimumPathMatrix(mapModuleIndex(m))(mapModuleIndex(dependency)) += dependency
    }
    applyFloydWarshallMainIteration(minimumPathMatrix, modulesSeq.size)
    new ModuleGraphTransitiveClosure(minimumPathMatrix, mapModuleIndex, modulesSeq)
  }

  private def applyFloydWarshallMainIteration(minimumPathMatrix: Array[Array[ArrayBuffer[HWModule]]], size: Int): Unit = {
    for {
      i <- 0 until size
      j <- 0 until size
      k <- 0 until size
    } {
      val minimumIJ = minimumPathMatrix(i)(j)
      val minimumIK = minimumPathMatrix(i)(k)
      val minimumKJ = minimumPathMatrix(k)(j)
      if (minimumIJ.isEmpty && minimumIK.nonEmpty && minimumKJ.nonEmpty
        || minimumIJ.nonEmpty && minimumIK.nonEmpty && minimumKJ.nonEmpty && (minimumIK.size + minimumKJ.size < minimumIJ.size)) {
        minimumPathMatrix(i)(j) = minimumIK ++ minimumKJ
      }
    }
  }
}
